lib/class-wp-annotation-utils.php

```diff
@@ -0,0 +1,820 @@
+		'FragmentSelector',
+		'CssSelector',
+		'XPathSelector',
+		'TextQuoteSelector',
+		'TextPositionSelector',
+		'DataPositionSelector',
+		// 'SvgSelector', Disabled for now.
+		'RangeSelector',
+	);
+
+	/**
+	 * Custom annotation statuses.
+	 *
+	 * @since [version]
+	 *
+	 * @var string[]
+	 */
+	public static $custom_statuses = array(
+		'resolve', // Archived as resolved.
+		'reject',  // Archived as rejected.
+		'archive', // Archived for another reason.
+	);
+
+	/**
+	 * Gets comment data.
+	 *
+	 * @since [version]
+	 *
+	 * @param  WP_Comment|int $comment Comment object or ID.
+	 *
+	 * @return stdClass|null           Data object. Null on failure.
+	 */
+	public static function get_comment_data( $comment ) {
+		if ( ! $comment ) {
+			return null;
+		}
+
+		$comment = get_comment( $comment );
+
+		if ( ! $comment ) {
+			return null;
+		}
+
+		$type        = $comment->comment_type;
+		$status      = self::get_comment_status( $comment );
+		$meta_status = self::get_comment_status( $comment, true );
+
+		if ( ! $type || ! $status || ! $meta_status ) {
+			return null;
+		}
+
+		if ( ! $comment->comment_post_ID ) {
+			return null;
+		}
+
+		$post_data = self::get_post_data( $comment->comment_post_ID );
+
+		if ( ! $post_data ) {
+			return null;
+		}
+
+		return (object) array(
+			'obj'         => $comment,
+			'type'        => $type,
+			'status'      => $status,
+			'meta_status' => $meta_status,
+			'post'        => $post_data,
+		);
+	}
+
+	/**
+	 * Gets a comment's parent post data.
+	 *
+	 * @since [version]
+	 *
+	 * @param  WP_Post|int $post Post object or ID.
+	 *
+	 * @return stdClass|null     Data object. Null on failure.
+	 */
+	public static function get_post_data( $post ) {
+		if ( ! $post ) {
+			return null;
+		}
+
+		$post = get_post( $post );
+
+		if ( ! $post ) {
+			return null;
+		}
+
+		if ( 'revision' === $post->post_type ) {
+			return null;
+		}
+
+		$type   = get_post_type_object( $post->post_type );
+		$status = get_post_status_object( get_post_status( $post ) );
+
+		if ( ! $type || ! $status ) {
+			return null;
+		}
+
+		return (object) array(
+			'obj'    => $post,
+			'type'   => $type,
+			'status' => $status,
+		);
+	}
+
+	/**
+	 * Gets comment status (supports custom statuses).
+	 *
+	 * Unlike {@see wp_get_comment_status()}, this supports custom statuses. Otherwise,
+	 * it closely resembles {@see wp_get_comment_status()}.
+	 *
+	 * @since [version]
+	 *
+	 * @param  WP_Comment|int $comment    Comment (i.e., annotation) object or ID.
+	 * @param  bool           $check_meta Optional. Default is false. If true, return
+	 *                                    would-be status if restored from spam or trash.
+	 *
+	 * @return string|bool                'approved', 'unapproved', 'spam', 'trash', or a
+	 *                                    custom status. Returns false on failure.
+	 *
+	 * @see wp_get_comment_status()
+	 */
+	public static function get_comment_status( $comment, $check_meta = false ) {
```

Comment:
If you have time to elaborate just a bit on why you feel this would cause more trouble than it's worth, I'd be curious to know what your experience has been; i.e., more about why you feel that way about `comment_approved`.

> I think these statuses should be put in comment meta. This also makes it easier to migrate to the main comments table if it ever gets a comment_status field.

Isn't that what the `comment_approved` column is already? I realize the name of this field is not ideal, but this is what the family of comment status handlers uses now; e.g., [`wp_set_comment_status()`](https://developer.wordpress.org/reference/functions/wp_set_comment_status/)

Reply:
Ok, you've changed my mind. It is fine to use it like this. Core needs first-class support for this in the future. But that could be done during merge proposal time I think.